packages/artifact/src/internal/upload/upload-artifact.ts

```diff
@@ -28,13 +31,74 @@ export async function uploadArtifact(
     }
   }
 
+  // get the IDs needed for the artifact creation
+  const backendIds = getBackendIdsFromToken()
+  if (!backendIds.workflowRunBackendId || !backendIds.workflowJobRunBackendId) {
+    core.warning(`Failed to get backend ids`)
+    return {
+      success: false
+    }
+  }
+
+  // create the artifact client
+  const artifactClient = createArtifactTwirpClient('upload')
+
+  // create the artifact
+  const createArtifactReq: CreateArtifactRequest = {
+    workflowRunBackendId: backendIds.workflowRunBackendId,
+    workflowJobRunBackendId: backendIds.workflowJobRunBackendId,
+    name,
+    version: 4
+  }
+
+  // if there is a retention period, add it to the request
+  const expiresAt = getExpiration(options?.retentionDays)
+  if (expiresAt) {
+    createArtifactReq.expiresAt = expiresAt
+  }
+
+  const createArtifactResp = await artifactClient.CreateArtifact(
+    createArtifactReq
+  )
+  if (!createArtifactResp.ok) {
+    core.warning(`Failed to create artifact`)
+    return {
+      success: false
+    }
+  }
+
   // TODO - Implement upload functionality
 
+  // finalize the artifact
+  const finalizeArtifactResp = await artifactClient.FinalizeArtifact({
+    workflowRunBackendId: backendIds.workflowRunBackendId,
+    workflowJobRunBackendId: backendIds.workflowJobRunBackendId,
+    name,
+    size: '0' // TODO - Add size
+  })
+  if (!finalizeArtifactResp.ok) {
+    core.warning(`Failed to finalize artifact`)
+    return {
+      success: false
+    }
+  }
+
   const uploadResponse: UploadResponse = {
     success: true,
     size: 0,
-    id: 0
+    id: parseInt(finalizeArtifactResp.artifactId) // TODO - will this be a problem due to the id being a bigint?
   }
 
   return uploadResponse
 }
+
+function getExpiration(retentionDays?: number): Timestamp | undefined {
```

Comment:
The expiration logic is a bit more complex.

We have to factor in any org/repo level settings that may be configured as that is the maximum allowed value. That is available via this env variable https://github.com/actions/toolkit/blob/91d3933eb52b351f437151400a88ba7d57442a9b/packages/artifact/src/internal/config-variables.ts#L65-L67

Here is this current bit of logic we have
https://github.com/actions/toolkit/blob/91d3933eb52b351f437151400a88ba7d57442a9b/packages/artifact/src/internal/upload-http-client.ts#L71C13-L75

https://github.com/actions/toolkit/blob/91d3933eb52b351f437151400a88ba7d57442a9b/packages/artifact/src/internal/utils.ts#L276-L295

Reply:
Ah, I was having trouble finding that logic! Thank you!